### PR TITLE
AUDIO Function

### PR DESCRIPTION
**Pull request by bjalder26**
This PR would allow sound clips to be played by Routines using an `AUDIO` function.

| parameters | description |
|-------------|-------------------------------|
| source | The url of the audio file. |
| type | Helps find the file, not required. |
| maxVolume | From 0 to 1. 1 is equal to 100%. The volume of the sound is a product of the volume set by the game creator times the volume of the slider.  The game creator sets the maxVolume, and the player can reduce the volume using the volume slider.  If a game creator sets the volume > 1, then it will automatically be set to 1 instead to minimize errors. |
| length | Length the audio clip should play in milliseconds.  Default/null will play full duration. |
| player | Can set a player that will be the only one that hears the sound.  Default/null means everyone hears the sound. |
 
Only 1 audio will play per widget's routine triggered.  If you want >1 audio to play at the same time, from a single event, then they will have to be in different widget's routines. 

Topics that should be considered:
- Volume control: currently the game maker sets a maximum volume for each sound to equalize the volume of different sounds, and the player can reduce the volume by using the volume slider.
- Should the maxVolume be on a scale of 0 to 100 instead of 0 to 1?  I used 0 to 1 because that is what is actually used in the [HTML DOM property](https://www.w3schools.com/tags/av_prop_volume.asp#:~:text=HTML%20Audio%2FVideo%20DOM%20volume%20Property%201%20Definition%20and,number%20between%200.0%20and%201.0.%205%20Technical%20Details).  However it won't increment in the JSON editor unless we eventually support incrementing decimals.

Proposed wiki changes:
In functions:
* [AUDIO](#audio) to play a sound

# AUDIO
This function causes a sound to play.  A url for the audio file is required.  Do not put this function inside of an audioChangeRoutine or it will cause an infinite loop.  It will not trigger any other changeRoutine.

Parameters:

- source: url - the url of the audio file.  It can play an audio clip from another website, or audio assets can be uploaded through the JSON editor.
- type:  audio/midi, audio/mpeg, audio/ogg, audio/wav, audio/x-wav - the type of the audio file.  While not required, it helps the program find the audio file.  MP3 files are of type audio/mpeg.
- maxVolume: number - the maximum volume is set by the game maker on a range from 0 to 1 where 1 is equal to 100%.  The actual volume played will be the product of the game maker's maxVolume setting and the volume setting chosen by the player with the volume slider.  The game maker can thus equalize the volume of all the sounds in their game, while the player still has some control over the overall volume.
- length: Length the audio clip should play in milliseconds.  Default/null will play full duration.
- player: playerID - can indicate which player will hear the sound.  Defaults to null which means every player hears the sound.

```json
    {
      "func": "AUDIO",
      "source": "https://website.org/sounds/soundfile.mp3",
      "type": "audio/mpeg",
      "volume": 1.0
      "player": "${playerName}"
    }
```
In widgets:
| Property  | Default value       | Description           | Tutorial              |
| ----- | ----- | ----- |-----|
| audio | null | Temporary property to trigger audio to play.  The `audio` property will not trigger a changeRoutine (to prevent infinite loops), but will trigger an audioChangeRoutine.  Do not try to use your own `audio` property on a widget, because the property will get deleted automatically. ||
| clickSound | null | Indicates the url of a sound to automatically play when the widget is clicked on.  Placing an AUDIO function in a clickRoutine will override this. ||

Source of icons:
https://fonts.google.com/icons?selected=Material%20Icons%3Avolume_up%3A
https://fonts.google.com/icons?selected=Material%20Icons%3Avolume_off%3A
https://fonts.google.com/icons?selected=Material%20Icons%3Asettings%3A

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:3608/pr-test (or any other room on that server)

**Comment by RaphaelAlvez** [comment]
It works.

I don't like the sound volume bar displayed like that. Personally I would add a configurations button and have the volume bar in that overlay.

Minotalen on discord suggested to me that we could use something like this https://www.bfxr.net/

So SOUND input would be a string of numbers to auto generate sound

**Comment by 96LawDawg** [comment]
I like it!  I'm going to disagree with Raphael on the volume slider.  I think it needs to be accessible all the time.  Not every player is going to want sound coming out and if the game designer has to put something else on screen, or the player has to go somewhere to tweak the sound, then that is not good.  There is plenty of space on that blue bar.  Use it.  You could even make a longer vertical slider to it is not so sensitive to scale.  But I'm not sure I like the grey because it changes the look too much.  Maybe have a white "🕪" (unicode speaker symbol in case you can't see it on your device) as a mute button that changes to something like that with a slash to show muted.  Then a white slider button on a light gray background.

**Comment by bjalder26** [comment]
> It works.
> 
> I don't like the sound volume bar displayed like that. Personally, I would add a configurations button and have the volume bar in that overlay.
> 
> Minotalen on discord suggested to me that we could use something like this https://www.bfxr.net/
> 
> So SOUND input would be a string of numbers to auto generate sound

I'm sorry, I can't see that link because it uses flash but I tried to look up some images.  As far as the slider, it doesn't really matter to me so much what we do.  However, it was a suggestion that the default should be muted.  If the default is muted then I think something should be on the bar to unmute it.  However, that could be something like LawDawg suggests below to mute and unmute on the bar, and another screen with options.  I didn't want to open the door to adding an options screen, but if that is something that is wanted, then it is fine with me.

My main goal was to get a function that would allow sound effects, and then to allow players control over the volume.

**Comment by bjalder26** [comment]
> I like it! I'm going to disagree with Raphael on the volume slider. I think it needs to be accessible all the time. Not every player is going to want sound coming out and if the game designer has to put something else on screen, or the player has to go somewhere to tweak the sound, then that is not good. There is plenty of space on that blue bar. Use it. You could even make a longer vertical slider to it is not so sensitive to scale. But I'm not sure I like the grey because it changes the look too much. Maybe have a white "🕪" (unicode speaker symbol in case you can't see it on your device) as a mute button that changes to something like that with a slash to show muted. Then a white slider button on a light gray background.

It could easily be vertical, and the scale is a percent of the length, so nothing would need to change there.  A speaker symbol sounds like a good idea to me.  I tried a checkbox, but didn't like how it looked.

I'm not sure how it appears on your screen.  This is how it appears on mine.  Left is during mouseover, and right is the rest of the time.
image

**Comment by 96LawDawg** [comment]
My image is like below when no mouseover.  On mouseover, it gets darker gray.

slider

Posted actual usable speaker icon suggested images in the pr test server.

**Comment by robartsd** [comment]
Slider inputs have very different styling on different browsers and do but have a standardized way to style. https://codepen.io/robartsd/pen/ExPywEL

**Comment by bjalder26** [comment]
> You could even make a longer vertical slider to it is not so sensitive to scale. 
I tried to make it vertical with disastrous results.  There are 3 different views as the bar rotates to go across the bottom and I can't get it to consistently wind up in the same spot on different browsers.  Maybe there is a way to do it, but I wouldn't know where to start.

**Comment by bjalder26** [comment]
I think a new widget would need to be created for the audio to be heard by all players.  No time to pursue it currently.

**Comment by RaphaelAlvez** [comment]
anyreason to come back at it?

File: client/css/layout.css [hunk 1]
```diff
@@ -109,8 +113,83 @@ button:hover {
...
+    margin-top: 0;
+    box-sizing: border-box;
+    border: none;
+    width: var(--thumbD);
+    height: var(--thumbD);
+    border-radius: 50%;
+    background: #2771ce;
+}
+[type='range']::-ms-tooltip {
+    display: none;
```
Comment (RaphaelAlvez) [inline client/css/layout.css:187]: you need to keep some of this css in a class. we have other sliders already

File: client/css/layout.css [hunk 2]
```diff
@@ -295,6 +381,14 @@ body.aspectTooGood #hideToolbarButton {
   background-image:url(/i/icons-white/info-circle-solid-svg.svg);
 }
 
+#muteButton {
+  background-image:url(/i/icons-white/mute-svg.svg);
+}
+
+#muteButton.toolbarButton.muted {
+  background-image:url(/i/icons-white/unmute-svg.svg);
+}
```
Comment (RaphaelAlvez) [inline client/css/layout.css:390]: I would rather use font awesome icons just to keep the current style. if you need any new icon I can work on it.

https://fontawesome.com/v5.15/icons/volume-up?style=solid
https://fontawesome.com/v5.15/icons/volume-mute?style=solid

**Review by RaphaelAlvez — CHANGES_REQUESTED** [review]
sliders are broken and icons are not in the current style

**Comment by 96LawDawg** [comment]
This should probably be approved as is.  I like the new volume slider.  It looks slightly different on Edge, Chrome, and Firefox, but all are fine.  A couple of observations (that shouldn't prevent this from being approved).

Lag is very noticeable for some of the clips.

Having multiple windows open at the same time playing the same clip gives awful results (I usually only do this in testing and muting one of the screens fixed that).

Having a button that clicked multiple widgets to make sound did not work.

The maxVolume parameter didn't seem to have much of an effect.

The only disaster I encountered was when I set up a short sound to play as a changeRoutine.  I slid the object around the room a few inches.  The sound played repeatedly for maybe 20 seconds (probably once for every 0.1 of a change in x or y), then stopped.  Then sound no longer works in the room and I got a strange thing in the JSON.  See:
changeRoutine

**Comment by bjalder26** [comment]
> This should probably be approved as is. I like the new volume slider. It looks slightly different on Edge, Chrome, and Firefox, but all are fine. A couple of observations (that shouldn't prevent this from being approved).

The slider uses the CSS for all sliders.  I know that in edge part of it appears grey, and if I can figure out how to change that I will, but probably as a separate PR.
 
> Lag is very noticeable for some of the clips.

Is it lag, or silence at the beginning of the clip?  Because I didn't edit any of the clips, and some of them have silence at the beginning.
 
> Having multiple windows open at the same time playing the same clip gives awful results (I usually only do this in testing and muting one of the screens fixed that).

A property needs to be passed to both screens, so I think it will be quicker on the originating screen.  I think the real question though would be if it is synced with other actions on that screen.  Hopefully it is.

> Having a button that clicked multiple widgets to make sound did not work.

Hmm.. I got multiple sounds to play with 2 calls and with 2 clicks.  As long as the audio functions are in two different widgets, and something isn't causing a crash it should work.

> The maxVolume parameter didn't seem to have much of an effect.

I must have broken it.  That needs to be fixed.  Thanks for finding that.  It's always the same, because it is using a default value instead of the set value.

> The only disaster I encountered was when I set up a short sound to play as a changeRoutine. I slid the object around the room a few inches. The sound played repeatedly for maybe 20 seconds (probably once for every 0.1 of a change in x or y), then stopped. Then sound no longer works in the room and I got a strange thing in the JSON. See: changeRoutine

If it is a changeRoutine (not an xChangeRoutine or yChangeRoutine), then it will cause an infinite loop because it changes a property on the widget (twice).  An xChangeRoutine or yChangeRoutine can cause up to 100 sounds to be opened at the same time, which can cause it's own problems.

Right now I have a 0.1 second delay between then one audio property can be written and when it is deleted allowing the next one to be written.  Perhaps that should be a parameter as well so that someone can set an xChangeRoutine that only opens a new sound after a specified time like however many seconds the audio clip lasts.  I can't really think of a use, but if that would be useful, then it should be pretty easy.

The audio property you're seeing in the image you posted is the information that is passed to each screen so that they will all play sound.  If it is not being deleted, then it is probably because the changeRoutine caused an infinite loop crashing the browser, so that the property doesn't get deleted after 0.1 sec (or perhaps it is just rewriting it still).  

Right now I don't have any good ideas on how to deal with the infinite loop problem with basic changeRoutines.
I also need to think about if I need to deal with the audio property not getting deleted.  Was the audio property eventually deleted by itself?

**Comment by 96LawDawg** [comment]
> Is it lag, or silence at the beginning of the clip? Because I didn't edit any of the clips, and some of them have silence at the beginning.

Probably just silence.  Definitely no lag with the short clips, but if some of the files (like the music) don't start immediately, it seemed like lag.  Probably not.
> 
> > Having multiple windows open at the same time playing the same clip gives awful results (I usually only do this in testing and muting one of the screens fixed that).
> 
> A property needs to be passed to both screens, so I think it will be quicker on the originating screen. I think the real question though would be if it is synced with other actions on that screen. Hopefully it is.

Nothing needs to be done here.  Anybody with multiple windows open should know to silence one of them.

> > Having a button that clicked multiple widgets to make sound did not work.
> 
> Hmm.. I got multiple sounds to play with 2 calls and with 2 clicks. As long as the audio functions are in two different widgets, and something isn't causing a crash it should work.

Yeah, works now.  Must have tried it in a messed up room.

> > The maxVolume parameter didn't seem to have much of an effect.
> 
> I must have broken it. That needs to be fixed. Thanks for finding that. It's always the same, because it is using a default value instead of the set value. (on edit: now fixed)

Fixed now.


> The audio property you're seeing in the image you posted is the information that is passed to each screen so that they will all play sound. If it is not being deleted, then it is probably because the changeRoutine caused an infinite loop crashing the browser, so that the property doesn't get deleted after 0.1 sec (or perhaps it is just rewriting it still).
> 
> Right now I don't have any good ideas on how to deal with the infinite loop problem with basic changeRoutines. (maybe I should make it so they can't play in widgets with changeRoutines, and put an error in the debug OR maybe audio property changes shouldn't ever trigger basic changeRoutines)
> 
> I also need to think about if I need to deal with the audio property not getting deleted. Was the audio property eventually deleted by itself?

I don't know that this is fixable.  I just thought maybe there could be some cool effect when sliding around.  )But I definitely don't want this to become the MySpace of the internet where everybody uses awful sounds for everything.)  Probably just need a warning in the wiki.

The audio property did not go away until I made the `changeRoutine` a `clickRoutine`.

**Review by 96LawDawg — APPROVED** [review]
Approved through testing.

**Comment by bjalder26** [comment]
> The audio property did not go away until I made the `changeRoutine` a `clickRoutine`.

Oh okay, then it is what I was seeing too.  It just kept triggering itself in an infinite loop.

It is probably not necessary to ever have an audio property trigger a changeRoutine (you could always trigger it with something else).  However, a changeRoutine in a widget that also triggers a sound might be needed at some point.  So maybe I should just block the audio property from triggering a changeRoutine.

**Comment by robartsd** [comment]
Can audio files be hosted on VTT server? This would be important for library games so that the server can be self contained.

**Comment by bjalder26** [comment]
> Can audio files be hosted on VTT server? This would be important for library games so that the server can be self contained.

Yes, they CAN.  You can upload files like you would upload a token, and use the link you would use if it was an image.  This PR is not set up to force that for every file, because I thought it would use up a lot of space.  However it CAN be done.

**Comment by robartsd** [comment]
> > Can audio files be hosted on VTT server? This would be important for library games so that the server can be self contained.
> 
> Yes, they CAN. You can upload files like you would upload a token, and use the link you would use if it was an image. This PR is not set up to force that for every file, because I thought it would use up a lot of space. However it CAN be done.

I see the "Dice Roll" button near the bottom in the pr-test room is using a local sound file.

**Comment by RaphaelAlvez** [comment]
I couldn't test it on my phone and pc yet but here are my thoughts

- if you want to add sound to the default widget maybe we should add a property "audio" that is called on normal click
     - we are going to add new game piece soon and it would be nice to add sound to them without needing to add a clickroutine ( this may be included in another PR for sure)
- why are we restricting it to 30 secs? this basically removes the possibility to have music playing to create a mood
- @ArnoldSmith86 should we host the files as we do for images?
- the slider may need some restyling in the future to be more visible but I like the popup

after my testing with my phone and PC I'll approve it

**Comment by ArnoldSmith86** [comment]
We should host the files.

**Comment by RaphaelAlvez** [comment]
> We should host the files.

then this involves adding an upload button

**Comment by ArnoldSmith86** [comment]
It involves more than that. Currently only a few filetypes are accepted:

```Javascript
      if(content[0] == 0xff)
        res.setHeader('Content-Type', 'image/jpeg');
      else if(content[0] == 0x89)
        res.setHeader('Content-Type', 'image/png');
      else if(content[0] == 0x3c)
        res.setHeader('Content-Type', 'image/svg+xml');
      else if(content[0] == 0x47)
        res.setHeader('Content-Type', 'image/gif');
      else if(content[0] == 0x52)
        res.setHeader('Content-Type', 'image/webp');
      else
        Logging.log(`WARNING: Unknown file type of asset ${req.params.name}`);
```

**Comment by bjalder26** [comment]
> I couldn't test it on my phone and pc yet but here are my thoughts
> 
> * if you want to add sound to the default widget maybe we should add a property "audio" that is called on normal click
>   
>   * we are going to add new game piece soon and it would be nice to add sound to them without needing to add a clickroutine ( this may be included in another PR for sure)

Currently I'm using a temporary "audio" property as a way to pass the data to every widget on every screen that needs to play sound, though that could be changed.  I suppose the default for the source could be the value of a property in the widget.  I'm not sure how the normal click should make that happen though.  Would it be a default clickRoutine?

> * why are we restricting it to 30 secs? this basically removes the possibility to have music playing to create a mood

10 seconds.  I was only trying to add sounds for routines.  I thought there was some resistance to the idea of long audio clips, though I don't recall that well anymore.  That can be easily changed.  I think I could have it "get" the length of the audio clip and use that instead if that is what is desired.  Or if a limit is desired, then we could just plug any number in.

**Comment by bjalder26** [comment]
> We should host the files.

I don't have a problem with that, but I figured it would take up a lot of space quickly, and was worried about that.  With that being said, I uploaded an audio file using the upload token button, then copied and pasted the link (from image) into an AUDIO function and it worked fine.

I'm not sure what uploading an audio file should 'look like'.  Left to my own devices I would probably add a second button like the upload token button to the add widget screen, but have it put the link in a different property instead (maybe the default property RaphaelAlvez is talking about).  I'm not sure if that is the best idea though.  I'm also not yet sure why using the upload token button worked given the code you posted.

**Comment by RaphaelAlvez** [comment]
> Would it be a default clickRoutine?

The click function tends to use the same function behind used to construct the equivalent functions in click routine but it goes directly to it in JS.
For example when you click a game piece it calls the same flip( ) function that is called by FLIP but it does not use the code in the routines

so for this we would add a something like this.set( "audio","clickSound") in click ( ) of widgets



> 10 seconds. I was only trying to add sounds for routines. I thought there was some resistance to the idea of long audio clips, though I don't recall that well anymore. That can be easily changed. I think I could have it "get" the length of the audio clip and use that instead if that is what is desired. Or if a limit is desired, then we could just plug any number in.

I can remember it either but I just liked one of the audio files in that room that was setting a very nice mood

**Comment by robartsd** [comment]
> I'm not sure what uploading an audio file should 'look like'. Left to my own devices I would probably add a second button like the upload token button to the add widget screen, but have it put the link in a different property instead (maybe the default property RaphaelAlvez is talking about). I'm not sure if that is the best idea though. I'm also not yet sure why using the upload token button worked given the code you posted.

IIRC the code @ArnoldSmith86 posted is executed on the server when the asset is requested. The server does not know this is an audio file, but apparently the client/script doesn't care that it is served without a correct MIME type (unlike CSS stylesheets which are ignored if the MIME type is not `text/css`).

**Comment by bjalder26** [comment]
> why are we restricting it to 30 secs? this basically removes the possibility to have music playing to create a mood

I just essentially removed that replacing it with a length parameter, and a default of null which will play the entire clip.  I hope that is what people want.

> > We should host the files.
> 
> then this involves adding an upload button

I figured out how to add a button in the JSON editor.  It shows up for Audio -> source and clickSound.
image

> IIRC the code @ArnoldSmith86 posted is executed on the server when the asset is requested. The server does not know this is an audio file, but apparently the client/script doesn't care that it is served without a correct MIME type (unlike CSS stylesheets which are ignored if the MIME type is not `text/css`).

Should that be "fixed" then, or just left alone?

**Review by 96LawDawg — APPROVED** [review]
Revised features tested and approved.  I think this is ready to merge.

**Pull request merged**